translate.lua

```diff
@@ -107,30 +110,35 @@ local function main()
             goldScoreTotal = goldScoreTotal + results[b].goldScore
             goldWordsTotal = goldWordsTotal + #goldBatch[b].words
           end
-
-          for n = 1, #results[b].preds do
-            local sentence = translator:buildOutput(results[b].preds[n])
-
-            if n == 1 then
-              outFile:write(sentence .. '\n')
-              predScoreTotal = predScoreTotal + results[b].preds[n].score
-              predWordsTotal = predWordsTotal + #results[b].preds[n].words
+          if opt.print_input_embeddings then
+            outFile:write(sentId, ' ', table.concat(results_table[b], " "), '\n')
+          else
+            for n = 1, #results[b].preds do
+              local sentence = translator:buildOutput(results[b].preds[n])
+
+              if n == 1 then
+                outFile:write(sentence .. '\n')
+                predScoreTotal = predScoreTotal + results[b].preds[n].score
+                predWordsTotal = predWordsTotal + #results[b].preds[n].words
+
+                if #results[b].preds > 1 then
+                  _G.logger:info('')
+                  _G.logger:info('BEST HYP:')
+                end
+              end
 
               if #results[b].preds > 1 then
-                _G.logger:info('')
-                _G.logger:info('BEST HYP:')
+                _G.logger:info("[%.2f] %s", results[b].preds[n].score, sentence)
+                if n ~= 1 then -- for the n == 1 case, we already wrote out the hypothesis above
+                  outFile:write(sentence .. '\n')
```

Comment:
There is no need to have special cases for `n == 1` and `n ~= 1`. Just call `write` at the beginning of the loop for all `n`.